src/components/SidePanels/TabPanels/Tabs.tsx

```diff
@@ -0,0 +1,178 @@
+  const { rootRef, tabTitle: tabText, tabIndex, panelKey, viewLength, children, active, panelWidth, transferTab, createNewPanel, setActiveTab } = props;
+  const [dragOverSide, setDragOverSide] = useState<DroppableSide | undefined>();
+  const tabRef = useRef<HTMLDivElement>();
+  const ghostTabRef = useRef<HTMLDivElement>();
+  const dragging = useRef(false);
+  const location = useRef({ panelKey, tabIndex });
+
+  useEffect(() => { location.current = { panelKey, tabIndex }; }, [panelKey, tabIndex]);
+
+  useDrag({
+    elementRef: tabRef,
+    onMouseDown(event) {
+      const { panelKey, tabIndex } = { ...location.current };
+
+      setActiveTab(panelKey, tabIndex);
+      rootRef.current?.append(ghostTabRef.current!);
+      const tab = tabRef.current!;
+      const page = rootRef.current!.getBoundingClientRect();
+      const bbox = tab.getBoundingClientRect();
+      const [x, y] = [event.pageX, event.pageY];
+      const [oX, oY] = [bbox.left - page.left, bbox.top - page.top];
+
+      return { x, y, oX, oY, panelKey, tabIndex };
+    },
+    onMouseMove(event, data) {
+      if (!data) return;
+      document.body.style.cursor = 'grabbing' ;
+
+      dragging.current = true;
+      const { x, y, oX, oY } = data;
+
+      ghostTabRef.current!.style.display = 'block';
+      ghostTabRef.current!.style.top = `${event.pageY - (y - oY)}px`;
+      ghostTabRef.current!.style.left = `${event.pageX - (x - oX)}px`;
+    },
+    onMouseUp(event, data) {
+      tabRef.current?.append(ghostTabRef.current!);
+      ghostTabRef.current!.style.display = 'none';
+      document.body.style.cursor = 'auto' ;
+
+      if (!data || !dragging.current) return;
+      dragging.current = false;
+      const { x, y, oX, oY, panelKey, tabIndex } = data;
+      const headerHeight = 32;
+      const [nX, nY] = [event.pageX - (x - oX), event.pageY - (y - oY)];
+      const left = nX < 0 ? 0 : nX;
+      const implementedHeight = nY - headerHeight;
+      const top = implementedHeight < 0 ?  0 : implementedHeight;
+      const droppedOver = document.elementFromPoint(event.clientX, event.clientY);
+      const isDropArea = determineDroppableArea(droppedOver as HTMLElement);
+
+      if (!isDropArea) createNewPanel(panelKey, tabIndex, left, top);
+      else {
+        const dropTarget = document.elementFromPoint(event.clientX, event.clientY);
+        const dropTargetId = dropTarget?.id;
+
+        if (!dropTargetId || !dropTargetId?.includes('droppable')) return;
+        const droppedOnIndices = dropTargetId.split('_');
+        const receivingPanel = droppedOnIndices[0];
+        const receivingTab = parseInt(droppedOnIndices[1]);
+        const dropSide = determineLeftOrRight(event, dropTarget as HTMLElement);
+
+        if ((tabIndex === receivingTab || receivingTab === viewLength) && panelKey === receivingPanel) return; 
+        dropSide && transferTab(tabIndex, panelKey, receivingPanel, receivingTab, dropSide);
+      }
+    },
+  }, [tabRef]);
```

Comment:
Is this a dependencies array? References are stable so if it is a dependencies array we would want to probably put nothing here.